brainiak/utils/kronecker_solvers.py

```diff
@@ -0,0 +1,314 @@
+import tensorflow as tf
+
+__all__ = [
+    "tf_kron_mult",
+    "tf_masked_triangular_solve",
+]
+
+
+def tf_solve_lower_triangular_kron(L, y):
+    """ Tensorflow function to solve L x = y
+    where L = kron(L[0], L[1] .. L[n-1])
+    and L[i] are the lower triangular matrices
+
+    Arguments
+    ---------
+    L : list of 2-D tensors
+        Each element of the list must be a tensorflow tensor and
+        must be a lower triangular matrix of dimension n_i x n_i
+
+    y : 1-D or 2-D tensor
+        Dimension (n_0*n_1*..n_(m-1)) x p
+
+    Returns
+    -------
+    x : 1-D or 2-D tensor
+        Dimension (n_0*n_1*..n_(m-1)) x p
+
+    """
+    n = len(L)
+    if n == 1:
+        return tf.linalg.triangular_solve(L[0], y)
+    else:
+        x = y
+        na = L[0].get_shape().as_list()[0]
+        n_list = tf.stack([tf.cast(tf.shape(input=mat)[0], dtype=tf.float64) for mat in L])
+        n_prod = tf.cast(tf.reduce_prod(input_tensor=n_list), dtype=tf.int32)
+        nb = tf.cast(n_prod/na, dtype=tf.int32)
+        col = tf.shape(input=x)[1]
+
+        for i in range(na):
+            xt, xinb, xina = tf.split(x, [i*nb, nb, (na-i-1)*nb], 0)
+            t = xinb / L[0][i, i]
+            xinb = tf_solve_lower_triangular_kron(L[1:], t)
+            xina = xina - tf.reshape(
+                tf.tile(tf.slice(L[0], [i+1, i], [na-i-1, 1]), [1, nb*col]),
+                [(na-i-1)*nb, col]) * \
+                tf.reshape(
+                  tf.tile(tf.reshape(t, [-1, 1]), [na-i-1, 1]),
+                  [(na-i-1)*nb, col])
+            x = tf.concat(axis=0, values=[xt, xinb, xina])
+
+        return x
+
+
+def tf_solve_upper_triangular_kron(L, y):
+    """ Tensorflow function to solve L^T x = y
+    where L = kron(L[0], L[1] .. L[n-1])
+    and L[i] are the lower triangular matrices
```

Comment:
Since the name indicates upper triangular matrix, should L be lower or upper?

Reply:
If I understand this correctly (@narayanan2004's code) this is correct: an upper triangular solve but the input shows up as lower triangular (meaning it solves `L^T x = y` for input `L` and `y` as in the docstring).